### PR TITLE
Automatic Generation of Test Cases for Incremental Static Analysis

### PR DESCRIPTION
**Pull request by J2000A**
# Introduction
Implementation of the **Test Automation for Incremental Analysis - TAIA** as part of the bachelor's thesis "*Automatic Generation of Test Cases for Incremental Static Analysis*". The script enables the user to create and run incremental tests for Goblint based on mutations generated on a single c file provided by the user. With this, you can create with ease thousands of incremental tests based on the `.c` files from the normal regression tests. This script requires some changes to the analyzer repository, which are implemented in [this pull request](https://github.com/goblint/analyzer/pull/1096).

# Project Structure
**TAIA-test-automation-for-incremental-analysis** Contains all components of the "Test Automation for Incremental Analysis - TAIA"
**├── README.md** Start point for using the "Test Automation for Incremental Analysis - TAIA"
**├── RUN.sh** Script for creating and running incremental tests based on a `.c` file
**├── RUN_BATCH.sh**  Script for creating and running incremental tests on a directory
**├── clang-mutations** Directory containing the clang-tidy checks needed for the mutation generation
**│  &nbsp;&nbsp;&nbsp; ├── MUTATIONS.md**
**│  &nbsp;&nbsp;&nbsp; ├── README.md**
**├── sample-files** Directory containing files you can use as input files for getting started
**│  &nbsp;&nbsp;&nbsp; ├── batch-ignore-files** Directory with ignore-files you want to use when running on all regression tests
**├── scripts** Directory with all scripts for the Mutation Generator
**│  &nbsp;&nbsp;&nbsp; ├── README.md** Information about the scripts

# Implementation
For creating and running incremental tests with the "Test Automation for Incremental Analysis - TAIA" the following steps are executed:

- Processes the options from the terminal and asks the user interactively for the inputs if the information was not yet provided by the terminal options (**run_cli.py**)
- Generate one mutation using the clang-tidy checks (**generate_mutations.py**)
- Generate all possible mutations and manage them (**generate_programs.py**)
- Add ``__goblint_check()``s to the programs (**add_check.py**)
- Annotate the generated checks (**add_check_annotations.py**)
- Based on the programs with the annotated check create the test files (**generate_tests.py**)
- Run the tests with the `update_suite.rb` script from the analyzer repository (**run_tests.py**)

### CODE REVIEW

File: incremental-test-generation/scripts/generate_programs.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+    # Do not disable warn asserts as then the update_suite ruby script can't check the annotations
+    params = re.sub(r'--disable [\'"]*warn\.assert[\'"]*', '', params)
+
+    # Do not change ana.race.direct-arithmetic as it requires json files as input
+    params = re.sub(r'--enable [\'"]*ana\.race\.direct-arithmetic[\'"]*', '', params)
+    params = re.sub(r'--disable [\'"]*ana\.race\.direct-arithmetic[\'"]*', '', params)
+
+    # Do use ana.thread.context.create-edges as it otherwise requires json files
+    params = re.sub(r'--disable [\'"]*ana\.thread\.context\.create-edges[\'"]*', '', params)
```
[the remark this answers is not in the document]
Comment (J2000A) [reply incremental-test-generation/scripts/generate_programs.py]: Indeed, you are correct. The issue wasn't actually with the options but elsewhere. The respective lines have been removed. The tests are now passing with these options.

File: incremental-test-generation/scripts/generate_programs.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+    # Do not privatization mutex-meet-tid as some of the information may get lost during incremental analysis
+    params = re.sub(r'--set [\'"]*ana\.base\.privatization[\'"]* [\'"]*mutex-meet-tid[\'"]*', '', params)
```
[the remark this answers is not in the document]
Comment (J2000A) [reply incremental-test-generation/scripts/generate_programs.py]: I will look further into this. Seems to be a problem with the update_suite.rb script. I keep you updated.

[the remark this answers is not in the document]
Comment (J2000A) [reply incremental-test-generation/scripts/generate_programs.py]: I looked more into this and it was actually a problem with the generation of the annotations for the goblint checks. I removed this line.

**Comment by jerhard** [comment]
Close this for now, as we currently are not working on the incremental analysis.

**Pull request closed without merging**